includes/Jobs/GenerateProductFeed.php

```diff
@@ -0,0 +1,112 @@
+<?php
+
+namespace SkyVerge\WooCommerce\Facebook\Jobs;
+
+use Automattic\WooCommerce\ActionSchedulerJobFramework\AbstractChainedJob;
+use Automattic\WooCommerce\ActionSchedulerJobFramework\Utilities\BatchQueryOffset;
+use Exception;
+use WC_Facebookcommerce;
+use WP_Query;
+
+/**
+ * Class GenerateProductFeed
+ *
+ * @since 2.5.0
+ */
+class GenerateProductFeed extends AbstractChainedJob {
+
+	use BatchQueryOffset;
+
+	/**
+	 * Called before starting the job.
+	 */
+	protected function handle_start() {
+		// Optionally override this method in child class.
+	}
+
+	/**
+	 * Called after the finishing the job.
+	 */
+	protected function handle_end() {
+		// Optionally override this method in child class.
+	}
+
+	/**
+	 * Get a set of items for the batch.
+	 *
+	 * NOTE: when using an OFFSET based query to retrieve items it's recommended to order by the item ID while
+	 * ASCENDING. This is so that any newly added items will not disrupt the query offset.
+	 *
+	 * @param int   $batch_number The batch number increments for each new batch in the job cycle.
+	 * @param array $args         The args for the job.
+	 *
+	 * @throws Exception On error. The failure will be logged by Action Scheduler and the job chain will stop.
+	 */
+	protected function get_items_for_batch( int $batch_number, array $args ): array {
```

Comment:
This approach has the potential to slow things down significantly and increase memory footprint. There is logic inside the `get_all_product_ids_for_sync` that prepares the item for batching later. I understand your concern about the size of the option which stores the items for processing but without it, you will be forced to do additional queries to figure out if the product variation that you are processing has a parent in the published state.

Reply:
Not according to my tests, I've got 20,000 products on my local site and this is what I observe:

Calling `get_all_product_ids_for_sync` and writing all the product IDs to an option:

<img width="1479" alt="Screen Shot 2021-04-30 at 9 29 30 am" src="https://user-images.githubusercontent.com/1892137/116635975-57087b80-a99f-11eb-9261-afd643923395.png">

Here I've used `wc_get_products()` (as you suggested in another comment) to select full product objects using an offset.

<img width="1490" alt="Screen Shot 2021-04-30 at 9 28 03 am" src="https://user-images.githubusercontent.com/1892137/116636041-7c958500-a99f-11eb-984f-80ff7610969c.png">

The actual code used:
```php

wc_get_products(
	[
		'status'  => 'publish',
		'type'    => [ 'simple', 'variation' ],
		'limit'   => 10,
		'offset'  => 0,
		'orderby' => 'ID',
		'order'   => 'ASC',
	]
);
```

Reply:
That is true but you do the query in `get_all_product_ids_for_sync`  only once and that only once and later you just read the option.
But the problem here is different. You don't know if the products that you have collected are valid! Variation in the `publish` state should not be synced unless its parent is also in publish state. So if you happen to have in your query 10 variation. You will need to figure out their parent id and query for their parent's state. With bigger batch numbers this could big a long process. This is exactly why `get_all_product_ids_for_sync` uses two separate queries and combines the results - this was the trick that allowed us to reduce ids collection time significantly.

Reply:
Gotcha, that's annoying that you can't query variations by status without referencing the parent product. 

I guess my concern is more about scalability rather than optimization. e.g. if a store has 100,000 products the option is going to be ~2 MB by my calculations. That is kind of an extreme amount of products but the more products a store has the high chance there is of the feed crashing with the options approach. If we use offset queries then the feed just takes longer to generate.

Reply:
What I am happy about is that the things that we discuss are interchangeable pieces. We could probably skip all of the discussions now and just push forward. When we will have most of this build we will just test. We will know what is the right approach without guessing.

Reply:
Yes that's true, we can adjust this later and work on other things.

But if we need to determine the variation status from the parent I think a JOIN is all we need.

```sql
SELECT 
	post.ID, 
	parent.ID as parentID,
	parent.post_status as status
FROM wp_posts as post
LEFT JOIN wp_posts as parent ON post.post_parent = parent.ID 
WHERE post.post_type = 'product_variation'
AND parent.post_status = 'publish'
ORDER BY post.ID ASC
LIMIT 10
``` 

I don't think there's any need to read and write big options values if we use a join like this.

Reply:
The easiest way to test it is to remove the limit and compare the total ids.

Reply:
We can actually select both the variations and the simple products in the same query...
```sql

SELECT post.ID
FROM wp_posts as post
LEFT JOIN wp_posts as parent ON post.post_parent = parent.ID 
WHERE 
	( post.post_type = 'product_variation' AND parent.post_status = 'publish' )
OR 
	( post.post_type = 'product' AND post.post_status = 'publish' )
ORDER BY post.ID ASC
LIMIT 50 OFFSET 0
```

Seems to perform quite well on my site

<img width="1267" alt="Screen Shot 2021-04-30 at 3 07 40 pm" src="https://user-images.githubusercontent.com/1892137/116653247-f04a8880-a9c5-11eb-90ba-a68997e157bd.png">